chainstate/tx-verifier/src/transaction_verifier/input_check/mod.rs

```diff
@@ -0,0 +1,458 @@
+    }
+
+    fn sub_ctx(&self) -> InputVerifyContextTimelock<'a, S> {
+        InputVerifyContextTimelock::new(self.ctx.sub_ctx, self.input_num)
+    }
+}
+
+impl<T, S: TransactionVerifierStorageRef> TimelockContext for InputVerifyContextFull<'_, T, S> {
+    type Error = ConnectTransactionError;
+
+    fn spending_height(&self) -> BlockHeight {
+        self.sub_ctx().spending_height()
+    }
+
+    fn spending_time(&self) -> BlockTimestamp {
+        self.sub_ctx().spending_time()
+    }
+
+    fn source_height(&self) -> Result<BlockHeight, Self::Error> {
+        self.sub_ctx().source_height()
+    }
+
+    fn source_time(&self) -> Result<BlockTimestamp, Self::Error> {
+        self.sub_ctx().source_time()
+    }
+}
+
+impl<T: Transactable, S> SignatureContext for InputVerifyContextFull<'_, T, S> {
+    type Tx = T;
+
+    fn chain_config(&self) -> &ChainConfig {
+        self.ctx.sub_ctx.chain_config
+    }
+
+    fn transaction(&self) -> &Self::Tx {
+        self.ctx.transaction
+    }
+
+    fn input_utxos(&self) -> &[Option<&TxOutput>] {
+        &self.ctx.spent_utxos
+    }
+
+    fn input_num(&self) -> usize {
+        self.input_num
+    }
+}
+
+pub trait FullyVerifiable<AV, TV>:
+    Transactable + for<'a> TranslateInput<TranslationContextFull<'a, &'a AV, &'a TV>>
+{
+}
+
+impl<T, AV, TV> FullyVerifiable<AV, TV> for T where
+    T: Transactable + for<'a> TranslateInput<TranslationContextFull<'a, &'a AV, &'a TV>>
+{
+}
+
+#[allow(clippy::too_many_arguments)]
+pub fn verify_full<T, S, UV, AV, TV>(
+    transaction: &T,
+    chain_config: &ChainConfig,
+    utxos_view: &UV,
+    pos_accounting: &AV,
+    tokens_accounting: &TV,
+    storage: &S,
+    tx_source: &TransactionSourceForConnect,
+    spending_time: BlockTimestamp,
+) -> Result<(), ConnectTransactionError>
+where
+    T: FullyVerifiable<AV, TV>,
+    S: TransactionVerifierStorageRef,
+    UV: utxo::UtxosView,
+    AV: pos_accounting::PoSAccountingView<Error = pos_accounting::Error>,
+    TV: tokens_accounting::TokensAccountingView<Error = tokens_accounting::Error>,
+{
+    let core_ctx = CoreContext::new(utxos_view, transaction)?;
+    let tl_ctx = VerifyContextTimelock::for_verifier(
+        chain_config,
+        storage,
+        tx_source,
+        spending_time,
+        &core_ctx,
+    );
+    let ctx = VerifyContextFull::new(transaction, &tl_ctx);
+
+    for (n, inp) in core_ctx.inputs_iter() {
+        let script =
+            TranslationContextFull::new(pos_accounting, tokens_accounting, inp).to_script::<T>()?;
+        let mut checker = mintscript::ScriptChecker::full(InputVerifyContextFull::new(&ctx, n));
+        script.verify(&mut checker)?;
+    }
+
+    Ok(())
+}
+
+pub fn verify_timelocks<T, S, UV>(
+    transaction: &T,
+    chain_config: &ChainConfig,
+    utxos_view: &UV,
+    storage: &S,
+    tip: Id<GenBlock>,
+    spending_height: BlockHeight,
+    spending_time: BlockTimestamp,
+) -> Result<(), ConnectTransactionError>
+where
```

Comment:
This doesn't feel to be in the correct place. The transaction verifier is supposed to be fully verifying transactions. Given that timelock constraints are only needed in the mempool, perhaps we should move this to the mempool. The reason why I'm thinking about this is that I don't want there to be any impression that the verification process is fractured.

Reply:
So, for simplicity, we agreed to add docs explaining this.